keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt

```diff
@@ -53,3 +74,99 @@ fun mapToArtifact(
     throw ArtifactParsingException(name, type, json, e)
   }
 }
+
+typealias ArtifactVersionSelectStep = SelectConditionStep<Record7<String, String, String, String, LocalDateTime, String, String>>
+typealias ArtifactVersionRow = ResultQuery<Record7<String, String, String, String, LocalDateTime, String, String>>
+
+internal fun ArtifactVersionRow.fetchArtifactVersions() =
+  fetch { (name, type, version, status, createdAt, gitMetadata, buildMetadata) ->
+    PublishedArtifact(
+      name = name,
+      type = type,
+      version = version,
+      status = status?.let { ArtifactStatus.valueOf(it) },
+      createdAt = createdAt?.toInstant(ZoneOffset.UTC),
+      gitMetadata = gitMetadata?.let { objectMapper.readValue(it) },
+      buildMetadata = buildMetadata?.let { objectMapper.readValue(it) },
+    )
+  }
+
+internal fun ArtifactVersionSelectStep.fetchSortedArtifactVersions(
```

Comment:
Utility function to avoid duplication when querying the database for artifact versions. This applies native SQL sorting and limiting when applicable, and falls back to the comparator from the sorting strategy if not.

Reply:
Thanks for the suggestion, @lorin. I'll break the function down. As for the last part of your comment, can you please explain what you mean by "side-effect style", maybe with an example?